ext/standard/tests/mail/bug80706.phpt

```diff
@@ -1,80 +1,57 @@
 --TEST--
-Bug #72964 (White space not unfolded for CC/Bcc headers)
---EXTENSIONS--
-imap
---CONFLICTS--
-imap
+Bug #80706 (Headers after Bcc headers may be ignored)
 --SKIPIF--
 <?php
-if (PHP_OS_FAMILY !== 'Windows') die('skip Windows only test');
 if (getenv("SKIP_SLOW_TESTS")) die('skip slow test');
-require_once __DIR__ . '/mail_skipif.inc';
+require_once __DIR__.'/mail_windows_skipif.inc';
 ?>
 --INI--
 SMTP=localhost
-smtp_port=25
+smtp_port=1025
+sendmail_from=from@example.com
 --FILE--
 <?php
-require_once __DIR__ . '/mail_include.inc';
 
-function find_and_delete_message($username, $subject) {
-    global $default_mailbox, $password;
+require_once __DIR__.'/mailpit_utils.inc';
 
-    $imap_stream = imap_open($default_mailbox, $username, $password);
-    if ($imap_stream === false) {
-        die("Cannot connect to IMAP server $server: " . imap_last_error() . "\n");
-    }
-
-    $found = false;
-    $repeat_count = 20; // we will repeat a max of 20 times
-    while (!$found && $repeat_count > 0) {
-        // sleep for a while to allow msg to be delivered
-        sleep(1);
-    
-        $num_messages = imap_check($imap_stream)->Nmsgs;
-        for ($i = $num_messages; $i > 0; $i--) {
-            $info = imap_headerinfo($imap_stream, $i);
-            if ($info->subject === $subject) {
-                $header = imap_fetchheader($imap_stream, $i);
-                echo "X-Mailer header found: ";
-                var_dump(strpos($header, 'X-Mailer: bug80706') !== false);
-                imap_delete($imap_stream, $i);
-                $found = true;
-                break;
-            }
-        }
-        $repeat_count--;
-    }
-
-    imap_close($imap_stream, CL_EXPUNGE);
-    return $found;
-}
-
-$to = "{$users[1]}@$domain";
+$to = 'bug80706_to@example.com';
+$from = ini_get('sendmail_from');
+$bcc = 'bug80706_bcc@exsample.com';
 $subject = bin2hex(random_bytes(16));
 $message = 'hello';
-$headers = "From: webmaster@example.com\r\n"
-    . "Bcc: {$users[2]}@$domain\r\n"
-    . "X-Mailer: bug80706";
+$xMailer = 'bug80706_x_mailer';
+$headers = "From: {$from}\r\n"
+    . "Bcc: {$bcc}\r\n"
+    . "X-Mailer: {$xMailer}";
 
 $res = mail($to, $subject, $message, $headers);
+
 if ($res !== true) {
-	die("TEST FAILED : Unable to send test email\n");
+    exit("Unable to send the email.\n");
 } else {
-	echo "Message sent OK\n";
+    echo "Sent the email.\n";
 }
 
-foreach ([$users[1], $users[2]] as $user) {
-    if (!find_and_delete_message("$user@$domain", $subject)) {
-        echo "TEST FAILED: email not delivered\n";
-    } else {
-        echo "TEST PASSED: Message sent and deleted OK\n";
+$res = searchEmailByToAddress($to);
+
+if (mailCheckResponse($res, $from, $to, $subject, $message)) {
+    echo "Received the email.\n";
+
+    $bccAddresses = getBccAddresses($res);
+    if (in_array($bcc, $bccAddresses, true)) {
+        echo "bcc Received the email.\n";
```

Comment:
NIT: This doesn't exactly tests that BCC received the email but more that BCC is set, right?

Reply:
yeah and might make sense to do the same for CC. Btw. shouldn't BCC be excluded from the original delivered email?

Reply:
Oops, I was careless.

mailpit (and mailhog) is like an SMTP dummy, so I may not be able to check the behavior on the receiving side.

This test checks "sent mail", not received mail.

Reply:
btw, it seems like it was tested like this before - checking inbox for each user. It doesn't check if BCC is present in the original though so that would be good to verify too to just see if there's bug. If there's a bug, it should just be logged and not fixed here - I'm just more interested if there's any issue atm.

Reply:
Properly delivering emails to multiple recipients and properly hiding BCC are all responsibilities of the SMTP server, so I think there is no need to see these when testing `mail()`.

Since `mail()` 's responsibility is to send mail data to the SMTP server, I think it is desirable to test whether the data received by the SMTP server from `mail()` is what we expected.

Reply:
Yes but we need to verify that data is sent correctly. In this case it means that multiple `RCPT TO` are sent which was possible in the previous version. You can see that the client have special handling for BCC in that regard: https://github.com/php/php-src/blob/5da833514533b4c8a6c95ece1f4af3e7b39107ca/win32/sendmail.c#L525-L614

The client actually strips BCC but seems that mailpit probably adds it back https://github.com/axllent/mailpit/commit/9c8329a05cef34b228a501cb14b3a9a346189905

If it's the case, it seems to me that mailpit might be good more for testing the actual email messages rather than making sure that client works as expected. For that it's maybe better to keep the actual real server and possibly just only replace imap client to check it out.

Reply:
That makes sense. It's already late today, so I'll try it tomorrow.

Reply:
I have done some checking but only CLI clients seem to be like complete CLI apps (e.g. mutt and alpine) but didn't find anything that would not be probably ideal.

But then I found also this project: https://github.com/nodemailer/wildduck which offers API. I guess it can work just with SMTP mailserver but we already got so maybe that would be a better option. We would basically just replace usage of imap ext with calls to this service that could check the same thing potentially.

Reply:
thank you. mailpit is a dummy SMTP server, so I may need something else to actually send.

If we want to test both sending and receiving behavior, things can get a little complicated.

I'll look into it a little more too.

Reply:
In previous tests, the sending/receiving server was the hmail server, and the IMAP extension was acting as an IMAP/POP3 client to the hmail server. wildduck apparently primarily functions as a receiving server, so it needs to be able to receive emails for tests, which can get quite complicated. (In fact, if I look at the documentation, there will be instructions to configure DNS.)

I'll continue investigating.